torchmultimodal/modules/layers/quantization.py

```diff
@@ -24,24 +25,61 @@ class Quantization(nn.Module):
     Vector quantization was introduced in Oord et al. 2017 (https://arxiv.org/pdf/1711.00937.pdf)
     to generate high-fidelity images, videos, and audio data.
 
+    The embedding weights are trained with exponential moving average updates as described
+    in original paper.
+
+    Code was largely inspired by a PyTorch implementation of the author's original code, found here:
+    https://colab.research.google.com/github/zalandoresearch/pytorch-vq-vae/blob/master/vq-vae.ipynb
+    and by the implementation in MUGEN (Hayes et al. 2022), found here:
+    https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/video_vqvae/vqvae.py
+
     Args:
         num_embeddings (int): the number of vectors in the embedding space
         embedding_dim (int): the dimensionality of the embedding vectors
 
     Inputs:
-        x (Tensor): Tensor containing a batch of encoder outputs.
+        z (Tensor): Tensor containing a batch of encoder outputs.
                     Expects dimensions to be batch x channel x n dims.
     """
 
-    def __init__(self, num_embeddings: int, embedding_dim: int):
+    def __init__(
+        self,
+        num_embeddings: int,
+        embedding_dim: int,
+        decay: float = 0.99,
+        epsilon: float = 1e-7,
+    ):
         super().__init__()
+        # Embedding weights and parameters for EMA update will be registered to buffer, as they
+        # will not be updated by the optimizer but are still model parameters.
+        # code_usage and code_avg correspond with N and m, respectively, from Oord et al.
+        self.register_buffer("embedding", torch.randn(num_embeddings, embedding_dim))
+        self.register_buffer("code_usage", torch.zeros(num_embeddings))
+        self.register_buffer("code_avg", self.embedding.clone())
+
         self.embedding_dim = embedding_dim
         self.num_embeddings = num_embeddings
 
-        self.embedding = nn.Embedding(self.num_embeddings, self.embedding_dim)
-        self.embedding.weight.data.uniform_(
-            -1 / self.num_embeddings, 1 / self.num_embeddings
-        )
+        self._decay = decay
+        # Used in Laplace smoothing of code usage
+        self._epsilon = epsilon
+
+        # Flag to track if we need to initialize embedding with encoder output
+        self._is_embedding_init = False
+
+    def _tile(self, x):
+        # Repeat encoder vectors in cases where the encoder output does not have enough vectors
+        # to initialize the codebook on first forward pass
+        num_encoder_vectors, num_channels = x.shape
+        if num_encoder_vectors < self.embedding_dim:
+            num_repeats = (
+                self.num_embeddings + num_encoder_vectors - 1
+            ) // num_encoder_vectors
+            # Add a small amount of noise to repeated vectors
+            std = 0.01 / torch.sqrt(num_channels)
```

Comment:
Is this value specific to the MUGEN implementation? Wondering if we could either generalize this implementation or explain why we're adding noise in this way

Reply:
This is originally from OpenAI, see [here](https://github.com/openai/jukebox/blob/08efbbc1d4ed1a3cef96e08a931944c8b4d63bb3/jukebox/vqvae/bottleneck.py#L28).

There's no mentioning of why this particular way of adding noise in their [paper](https://arxiv.org/pdf/2005.00341.pdf) but the logic looks generic to me:
1) Adding noise because we don't want the codebook vectors (constructed by repeating `x`) to look identical
2) `torch.rand_like(x)`: the noise follows a uniform distribution due to assumptions of VQVAE model
3) `* std`: normalize the noise along each codebook vector to be `U[0,1]`.

Reply:
Thanks for the explanation. So I guess I am more interested in why the value of 0.01 is being chosen here. It looks to me like this is actually a `N(0, 1e-4 * I)` distribution (modulo me probably screwing up some normalization). But this would be less of a standard (no pun intended) distribution. Lmk if I'm misunderstanding the implementation though

Reply:
I think the decision for 0.01 was mainly arbitrary to control the dispersion or magnitude of the noise. If it was true normal noise `N(0,1)` then that could change the encoder output too significantly, better to change it by `+/- 1e-4 / sqrt(num_channels)` most of the time instead

Reply:
Should 0.01 be a parameter then?